include/exec/materialize.hpp

```diff
@@ -0,0 +1,209 @@
+ * the License. You may obtain a copy of the License at
+ *
+ *   https://llvm.org/LICENSE.txt
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#pragma once
+
+#include <stdexec/execution.hpp>
+
+namespace exec {
+  namespace __materialize {
+    using namespace stdexec;
+
+    template <class _ReceiverId>
+    struct __receiver {
+      using _Receiver = stdexec::__t<_ReceiverId>;
+
+      class __t {
+       public:
+        __t(_Receiver&& __upstream)
+          : __upstream_{(_Receiver&&) __upstream} {
+        }
+
+       private:
+        _Receiver __upstream_;
+
+        template <__completion_tag _Tag, __decays_to<__t> _Self, class... _Args>
+          requires tag_invocable<set_value_t, _Receiver&&, _Tag, _Args...>
+        friend void tag_invoke(_Tag tag, _Self&& __self, _Args&&... __args) noexcept {
+          set_value((_Receiver&&) __self.__upstream_, tag, (_Args&&) __args...);
+        }
+
+        template <__decays_to<__t> _Self>
+        friend env_of_t<_Receiver> tag_invoke(get_env_t, _Self&& __self) noexcept {
+          return get_env(__self.__upstream_);
+        }
+      };
+    };
+
+    template <class _SenderId>
+    struct __sender {
+      using _Sender = decay_t<stdexec::__t<_SenderId>>;
+
+      template <class _Receiver>
+      using __receiver_t = stdexec::__t<__receiver<__id<decay_t<_Receiver>>>>;
+
+      class __t {
+       public:
+        using is_sender = void;
+
+        template <__decays_to<_Sender> _Sndr>
+        __t(_Sndr&& __sender)
+          : __sender_{(_Sndr&&) __sender} {
+        }
+
+        //  private:
+        _Sender __sender_;
+
+        template <__decays_to<__t> _Self, class _Receiver>
+          requires sender_to<__copy_cvref_t<_Self, _Sender>, __receiver_t<_Receiver>>
+        friend connect_result_t<_Sender, __receiver_t<_Receiver>>
+          tag_invoke(connect_t, _Self&& __self, _Receiver&& __receiver) noexcept(
+            __nothrow_connectable<__copy_cvref_t<_Self, _Sender>, __receiver_t<_Receiver>>) {
+          return connect(
+            ((_Self&&) __self).__sender_, __receiver_t<_Receiver>{(_Receiver&&) __receiver});
+        }
+
+        template <class... _Args>
+        using __materialize_value = completion_signatures<set_value_t(set_value_t, _Args...)>;
+
+        template <class _E>
+        using __materialize_error = completion_signatures<set_value_t(set_error_t, _E)>;
+
+        template <class _Env>
+        using __completion_signatures_for_t = make_completion_signatures<
+          _Sender,
+          _Env,
+          completion_signatures<>,
+          __materialize_value,
+          __materialize_error,
+          completion_signatures<set_value_t(set_stopped_t)>>;
```

Comment:
Add this one only if the input sender can complete with stopped.

Reply:
But isnt that what automatically happens? I am testing this

```cpp
  auto just_ = materialize(just());
  static_assert(sender<decltype(just_)>);
  static_assert(__v<__equivalent<
                  completion_signatures_of_t<decltype(just_)>,
                  completion_signatures<set_value_t(set_value_t)>>>);
```

I mean, I can add 

```cpp
          __if_c<
            sends_stopped<_Sender, _Env>,
            completion_signatures<set_value_t(set_stopped_t)>,
            completion_signatures<>>>
```

But I don't understand the rationale

Reply:
Oh yeah I forgot